[6.2][Concurrency] Fix nonisolated(nonsending) interaction with #isolation

ktoso opened the pull request:
**Description**: The `#isolation` did not work correctly with `nonisolated(nonsending)`, as that would add an implicit isolated parameter in SIL, yet the macro was only considering the AST parameters. This made it impossible to treat `nonisolated(nonsending)` as a "true" replacement for passing `isolated (any Actor)? = #isolation` to methods, because we'd not be able to "carry forward" the `#isolation` inside method bodies.

**Scope/Impact**: This affects users moving to `nonisolated(nonsending)` from passing explicit isolated parameters, and allows them to correctly "stay on" the caller as long as they want.
**Risk:** Low, this only corrects the macro expansion in SIL specifically for `nonisolated(nonsending)` methods.
**Testing**: CI testing, added runtime and SIL test.
**Reviewed by**: @DougGregor 

**Original PR:**  https://github.com/swiftlang/swift/pull/82793
**Radar:** rdar://155003540

ktoso committed:
[Concurrency] Fix nonisolated(nonsending) interaction with #isolation

ktoso committed:
[Concurrency] #isolation+nonsending nonisolated - review followup and SIL test

ktoso commented:
@swift-ci please test

jamieQ commented on test/Concurrency/Runtime/isolated_macro_in_nonisolated_nonsending_func.swift:
```diff
@@ -0,0 +1,72 @@
+  MainActor.assertIsolated()
+
+  let outerIsolation = #isolation
+  print("outerIsolation = #isolation = \(String(describing: outerIsolation))") // WRONG; this is nil today
```
nit: is this comment still relevant?

ktoso replied:
```suggestion
  print("outerIsolation = #isolation = \(String(describing: outerIsolation))")
```

jamieQ commented on test/Concurrency/Runtime/isolated_macro_in_nonisolated_nonsending_func.swift:
```diff
@@ -0,0 +1,72 @@
+    let iso = #isolation
+    print("Task{ [outerIsolation] } outerIsolation = \(String(describing: outerIsolation)), #iso = \(String(describing: iso))")
+  }.value
+
```
do you think it'd be worth calling a function that takes a defaulted isolated parameter within this `nonisolated(nonsending)` method to ensure the 'plumbing' of the isolation works as expected, or is that already covered elsewhere?

ktoso replied:
Default values are not really special here

ktoso reviewed (commented):
One comment fix up

ktoso committed:
Remove unnecessary comment in test

slavapestov commented on lib/SILGen/SILGenExpr.cpp:
```diff
@@ -7287,19 +7287,34 @@ RValue RValueEmitter::visitMacroExpansionExpr(MacroExpansionExpr *E,
+
+    if (isolation == ActorIsolation::CallerIsolationInheriting) {
+      auto *isolatedArg = SGF.F.maybeGetIsolatedArgument();
+      assert(isolatedArg &&
```
The following line is going to crash if the assert is skipped, so you may as well use ASSERT() instead

ktoso committed:
[Concurrency] Use ASSERT for to also assert in release builds

Small cleanup from code review; we'd crash either way after this line,
so might be better to always ASSERT more nicely.
